build/rules.mak

```diff
@@ -46,84 +22,77 @@ $(foreach d,$(MKDIRECTORIES),$(eval $(call mkdir_rule,$(d))))
-
-$(OUTPUT_DIR)/README.rst: $(ROOT_DIR)/README.rst
-	$(call trace_to_console, "Copying",$@)
-	$(_hide_cmds)$(call log_command,cp $< $@)
-
 $(OUTPUT_DIR)/setup.py: $(TEMPLATE_DIR)/setup.py.mako
 	$(call trace_to_console, "Generating",$@)
 	$(_hide_cmds)$(call log_command,$(call GENERATE_SCRIPT, $<, $(dir $@), $(METADATA_DIR)))
 
-sdist: $(SDIST_DONE)
+sdist: $(SDIST_BUILD_DONE) $(UNIT_TEST_FILES)
 
-$(SDIST_DONE): $(OUTPUT_DIR)/setup.py $(OUTPUT_DIR)/README.rst $(MODULE_FILES) $(UNIT_TESTS_PASSED)
+$(SDIST_BUILD_DONE): $(OUTPUT_DIR)/setup.py $(OUTPUT_DIR)/README.rst $(MODULE_FILES) $(UNIT_TESTS_PASSED)
 	$(call trace_to_console, "Creating sdist",$(OUTPUT_DIR)/dist)
 	$(_hide_cmds)$(call make_with_tracking_file,$@,cd $(OUTPUT_DIR) && python3 setup.py sdist $(LOG_OUTPUT) $(LOG_DIR)/sdist.log)
 
-wheel: $(WHEEL_DONE)
+wheel: $(WHEEL_BUILD_DONE) $(UNIT_TEST_FILES)
 
-$(WHEEL_DONE): $(OUTPUT_DIR)/setup.py $(OUTPUT_DIR)/README.rst $(MODULE_FILES) $(UNIT_TESTS_PASSED)
+$(WHEEL_BUILD_DONE): $(OUTPUT_DIR)/setup.py $(OUTPUT_DIR)/README.rst $(MODULE_FILES) $(UNIT_TESTS_PASSED)
 	$(call trace_to_console, "Creating wheel",$(OUTPUT_DIR)/dist)
 	$(_hide_cmds)$(call make_with_tracking_file,$@,cd $(OUTPUT_DIR) && python3 setup.py bdist_wheel --universal $(LOG_OUTPUT) $(LOG_DIR)/wheel.log)
 
-# From https://stackoverflow.com/questions/16467718/how-to-print-out-a-variable-in-makefile
-print-%: ; $(info $(DRIVER): $* is $(flavor $*) variable set to [$($*)]) @true
-
-$(TOX_INI): $(ROOT_DIR)/tox.ini
+$(OUTPUT_DIR)/README.rst: $(ROOT_DIR)/README.rst
 	$(call trace_to_console, "Copying",$@)
 	$(_hide_cmds)$(call log_command,cp $< $@)
 
-test: $(TOX_INI)
-	$(call trace_to_console, "Running tox",$(OUTPUT_DIR))
-	$(_hide_cmds)$(call log_command,cd $(OUTPUT_DIR) && set DRIVER=$(DRIVER) && tox)
+# From https://stackoverflow.com/questions/16467718/how-to-print-out-a-variable-in-makefile
+print-%: ; $(info $(DRIVER): $* is $(flavor $*) variable set to [$($*)]) @true
 
-update_generated_files: $(GENERATED_FILES_DONE)
+update_generated_files: $(GENERATED_FILES_COPY_DONE)
 
-# Can't use make_with_tracking_file since there are multiple commands
-$(GENERATED_FILES_DONE): $(MODULE_FILES) $(OUTPUT_DIR)/setup.py
+$(GENERATED_FILES_COPY_DONE): $(MODULE_FILES) $(OUTPUT_DIR)/setup.py $(UNIT_TEST_FILES)
 	$(call trace_to_console, "Updating",$(GENERATED_DIR)/$(DRIVER)/)
-	$(_hide_cmds)$(call log_command,touch $@)
-	$(_hide_cmds)$(call log_command,rm $@)
-	$(_hide_cmds)$(call log_command,rm -Rf $(GENERATED_DIR)/$(DRIVER))
-	$(_hide_cmds)$(call log_command,mkdir -p $(GENERATED_DIR)/$(DRIVER))
-	$(_hide_cmds)$(call log_command,cp -Rf $(MODULE_DIR)/* $(GENERATED_DIR)/$(DRIVER))
-	$(_hide_cmds)$(call log_command,cp -Rf $(OUTPUT_DIR)/setup.py $(GENERATED_DIR)/$(DRIVER))
-	$(_hide_cmds)$(call log_command,touch $@)
+	$(_hide_cmds)$(call make_with_tracking_file, $@, \
+      rm -Rf $(GENERATED_DIR)/$(DRIVER) && \
```

Comment:
this is pre-existing... but

if you rm dir/* instead, then the folder is emptied but you don't need to mkdir it right after...

Reply:
Changed. Also had to add to MKDIRS to ensure it existed